differ/windowsdiffer/differ_windows.go

```diff
@@ -0,0 +1,420 @@
+	}
+
+	cw, err := s.store.Writer(ctx, ref, 0, "")
+	if err != nil {
+		return emptyDesc, errors.Wrap(err, "failed to open writer")
+	}
+
+	layer, parentLayerPaths, err := mountsToLayerAndParents(upper)
+	if err != nil {
+		return emptyDesc, errors.Wrap(err, "failed to get layer and parent paths from mounts")
+	}
+	if lower[0].Source != parentLayerPaths[0] {
+		return emptyDesc, errors.Wrapf(errdefs.ErrInvalidArgument, "lower mounts must be the direct child of the upper mounts %v != %v", lower[0].Source, parentLayerPaths[0])
+	}
+	var opts []content.Opt
+	if isCompressed {
+		dgstr := digest.SHA256.Digester()
+		compressed, err := compression.CompressStream(cw, compression.Gzip)
+		if err != nil {
+			return emptyDesc, errors.Wrap(err, "failed to get compressed stream")
+		}
+		err = s.writeDiff(ctx, io.MultiWriter(compressed, dgstr.Hash()), layer, parentLayerPaths)
+		compressed.Close()
+		if err != nil {
+			return emptyDesc, errors.Wrap(err, "failed to write compressed diff")
+		}
+		opts = append(opts, content.WithLabels(map[string]string{
+			"containerd.io/uncompressed": dgstr.Digest().String(),
+		}))
+	} else {
+		if err = s.writeDiff(ctx, cw, layer, parentLayerPaths); err != nil {
+			return emptyDesc, errors.Wrap(err, "failed to write diff")
+		}
+	}
+
+	dgst := cw.Digest()
+	if err := cw.Commit(ctx, 0, dgst, opts...); err != nil {
+		return emptyDesc, errors.Wrap(err, "failed to commit")
+	}
+
+	info, err := s.store.Info(ctx, dgst)
+	if err != nil {
+		return emptyDesc, errors.Wrap(err, "failed to get info from content store")
+	}
+
+	return ocispec.Descriptor{
+		MediaType: media,
+		Size:      info.Size,
+		Digest:    info.Digest,
+	}, nil
+}
+
+func mountsToLayerAndParents(mounts []mount.Mount) (string, []string, error) {
+	if len(mounts) == 0 {
+		return "", nil, errors.Wrap(errdefs.ErrInvalidArgument, "number of mounts should not be 0")
+	}
+	layer := mounts[0].Source
+
+	var parents []string
+	for _, mount := range mounts[1:] {
+		parents = append(parents, mount.Source)
+	}
+
+	return layer, parents, nil
+}
+
+// WriteDiff writes a tar stream of the computed difference between the
+// provided directories.
+//
+// Produces a tar using OCI style file markers for deletions. Deleted
+// files will be prepended with the prefix ".wh.". This style is
+// based off AUFS whiteouts.
+// See https://github.com/opencontainers/image-spec/blob/master/layer.md
+func (s *windowsDiff) writeDiff(ctx context.Context, w io.Writer, layer string, parentLayerPaths []string) error {
+	info := hcsshim.DriverInfo{
+		Flavour: 1,
+		HomeDir: filepath.Dir(layer),
+	}
+	id := filepath.Base(layer)
+	err := winio.RunWithPrivilege(winio.SeBackupPrivilege, func() error {
+		r, err := hcsshim.NewLayerReader(info, id, parentLayerPaths)
+		if err != nil {
+			return err
+		}
+
+		err = writeTarFromLayer(r, w)
+		cerr := r.Close()
+		if err == nil {
+			err = cerr
+		}
+		return err
+	})
+	return err
+}
+
+const (
```

Comment:
move at top of file?

Reply:
This is going to be moved soon. It's only in the middle of the file as POC